[IV-22-23] Objetivo 2

jmramirezG opened the pull request:
## Sobre la estructura del repositorio

* [X] ¿He seguido las mejores prácticas en nombre de las clases y ficheros y
   disposición de los mismos?

## Sobre el análisis del problema

* [x] ¿Se ha documentado qué análisis se ha hecho sobre el dominio para decir lo
   que se ha creado?
* [x] ¿Se ha documentado por qué se ha elegido que lo creado sea un objeto valor,
   una entidad o un agregado?
* [x] ¿He propuesto un producto mínimamente viable, que en muchos casos será un solo
   objeto valor que no dependa de ningún otro (y que sea la base de muchos
   otros)?.

## Sobre la planificación y la programación

* [X] ¿Los issues representan un problema, y no una tarea?
* [X] ¿Todos los mensajes de commit explican el cambio, y no se
      limitan a repetir el nombre del fichero que se ha cambiado?
* [X] ¿Los mensajes de commit siguen el formato estándar y buenas prácticas?
* [x] ¿Se ha hecho una revisión real del código para comprobar que todos
   los atributos y funciones creadas están respaldadas por una HU?
* [x] ¿Todos los issues creados están asignados a una HU?
* [x] ¿Todos los cambios en el código están asignados a un issue al que se
      referencia en un commit? Los issues sobre los que estoy trabajando, ¿han
      sido asignados al primer milestone?
* [x] ¿Se ha asignado al mismo milestone el PR que se ha hecho?
* [X] ¿Son los milestones sobre los que estoy trabajando productos
   mínimamente viables? ¿O tengo que solicitar al product manager que
   cree milestones adicionales o precise de qué producto se trata?

jmramirezG committed:
Initialized go repo

jmramirezG commented:
Cierro issue #18, ya que hemos inicializado el repo con el lenguaje de programación escogido.

jmramirezG commented:
Puedes referenciar la milestone correspondiente en el PR? @iscoct

iscoct commented:
Está bien por ahora, espero a que se hagan los tipos de datos que se comentaron [aquí](https://github.com/iscoct/cotan/issues/19) para poder proponer el mergeo en el proyecto de JJ, ¿okay?

Todos los cambios necesarios para cumplir el objetivo 2 y el milestone deben de hacerse a la vez para ponerlos allí.

De todas formas es un buen comienzo, ánimo.

jmramirezG committed:
Adds data types

jmramirezG commented:
Avanza issue #19 porque declaramos estructuras de datos.

Hola @iscoct!

He subido los tipos de dato que se mencionan en el [M1](https://github.com/iscoct/cotan/milestone/4), teniendo en cuenta lo comentado en la issue #19.

Si quieres algún cambio ve comentándome!

iscoct commented on types/models/plato.go:
```diff
@@ -0,0 +1,53 @@
+type TipoPlato int64
+
+const (
+	Postre TipoPlato = iota
```
Creo que sería más fácil implementar la función String de esta manera:

func (s TipoPlato) String() string {
    return []string{"postre", "entrante", "primero", "segundo"}[s]
}

Por defecto iota empieza en 0, así que eso funcionaría.

Otra opción sería hacerlo así:

type TipoPlato string

const (
   Postre TipoPlato = "postre"
   Entrante = "entrante"
   Primero = "primero"
   Segundo = "segundo"
)

Mira a ver si puedes hacerlo de una de las 2 maneras si lo ves más sencillo también.

iscoct replied:
Lo mismo pasaría para EstadoPlato

jmramirezG replied:
He obtado por la segunda opción, si nos remitimos a [este enlace](https://www.sohamkamani.com/golang/enums/#enum-types) podemos ver que definir el tipo nos permite que TipoPlato (o EstadoPlato) dejen de ser comparables con strings, ya que tendrían tipos distintos.

iscoct replied:
Guay!

iscoct commented on iv.yaml:
```diff
@@ -0,0 +1,2 @@
+lenguaje: go
+entidad: main.go
```
En entidad debemos de poner las entidades que hemos creado en el proyecto.

Creo que en este caso puede ser entidad el tipo comanda porque gestiona los objetos Plato y Bebida.

[En la documentación del objetivo 2](http://jj.github.io/IV/documentos/proyecto/2.Entidad) viene descrito una entidad como: _clases que van a incluir lógica de negocio, y van a ser capaces de gestionar otras clases, así como almacenamiento y otros _objetos__ por lo que puede creo que puede ser esa.

¿Cómo lo ves?

jmramirezG replied:
He cambiado la entidad a la propuesta

iscoct replied:
Perfecto!

jmramirezG committed:
issue #19, change enums to string type

JJ commented on README.md:
```diff
@@ -27,4 +27,5 @@ Tiempo medio de comunicación del staff (camareros, pinches, chefs, ...) y prior
 - [Contexto](./docs/contexto.md): Se describe el contexto del problema, información sobre el sector de la gastronomía y beneficios de nuestro sistema.
 - [Personas](./docs/personas.md)
 - [Planificación](./docs/planificacion.md)
+  [Intalar y ejecutar](./docs/instalar.md)
```
A partir de este objetivo, en el README.md tienen que estar sólo las instrucciones para instalar y ejecutar. El resto pueden enlazarse desde el README.md de docs.

iscoct replied:
Está resuelto ya

JJ commented on main.go:
```diff
@@ -0,0 +1,5 @@
+package main
```
Nunca hemos pedido un main. Los productos deben ser mínimos.

iscoct replied:
Ya se ha eliminado con los nuevos cambios

JJ commented on types/models/plato.go:
```diff
@@ -0,0 +1,25 @@
+
+const (
+	Postre TipoPlato = "postre"
+	Entrante 		 = "entrante"
```
¿Qué relación tiene con la lógica de negocio?

iscoct replied:
Aquí hemos intentado explicar la relación con la lógica de negocio: https://github.com/iscoct/cotan/issues/25

JJ commented on types/models/mesa.go:
```diff
@@ -0,0 +1,5 @@
+package models
+
+type Mesa struct {
+	Numero int
```
¿Qué importancia tiene para la lógica de negocio el número de la mesa donde se encargue? ¿De qué forma es un producto mínimo?

iscoct replied:
En este issue hemos intentado explicar la importancia con la lógica de negocio: https://github.com/iscoct/cotan/issues/24

JJ commented on types/models/plato.go:
```diff
@@ -0,0 +1,25 @@
+type EstadoPlato string
+
+const (
+	EnEspera EstadoPlato = "en espera"
```
¿Qué relación tiene con la lógica de negocio?

iscoct replied:
En este issue: https://github.com/iscoct/cotan/issues/25 hemos intentado poner información al respecto con la relación de la lógica de negocio.

JJ reviewed (changes requested):
Un proyecto informático resuelve un problema y es un problema específico que está codificado en la historias de usuario. Por favor, desarrollad a partir de las historias de usuario como se ha indicado repetidamente en los objetivos semanales. A @iscoct, deja sólo las historias de usuario que lo sean de verdad (esencialmente, sólo #17). A @jmramirezg, crea issues no con "estructuras propuestas" sin ningúna relación con el problema sino con los problemas que se plantean a partir de la lógica de negocio.
Aparte, yo diría que aparentemente sólo habéis planteado objetos valor. No hay realmente ninguna entidad. En general, no se ha aplicado ningún aspecto del diseño dirigido por dominio.
Finalmente, os aconsejaría que hiciérais esto lo más rápidamente posible, porque es conveniente empezar pronto con el objetivo 3.

jmramirezG committed:
Remove main

jmramirezG committed:
Remove install and run instructions from docs

iscoct commented on docs/README.md:
```diff
@@ -0,0 +1,31 @@
+# Optimizador de Restaurantes
```
¿Por qué se ha puesto un README.md en docs?

Que yo sepa sólo debería de haber un README.md en el proyecto principal y no más de uno.

A lo mejor tiene sentido varios si una carpeta representa una parte grande del repositorio o proyecto, pero no creo que sea este el caso.

A lo mejor cuando me lo expliques tiene sentido, vamos hablando.

jmramirezG replied:
Como indicó JJ previamente:

>A partir de este objetivo, en el README.md tienen que estar sólo las instrucciones para instalar y ejecutar. El resto pueden enlazarse desde el README.md de docs.

Entonces he dejado en el README sólo las intrucciones de instalar y ejecutar.

iscoct replied:
Ah, guay @jmramirezG, no lo había leído bien, gracias por la corrección.

jmramirezG committed:
close #23 as we create Bebida datatype

jmramirezG committed:
close #23 as we create Bebida datatype

jmramirezG committed:
close #25 as we create Plato datatype

jmramirezG committed:
close #27 as we create Comanda datatype

jmramirezG committed:
close #24 as we create Mesa datatype

jmramirezG committed:
close #26 as we create Empleado datatype

iscoct commented:
Está listo para revisar por nuestra parte @JJ

JJ commented on types/models/plato.go:
```diff
@@ -0,0 +1,25 @@
+	Entregado 			 = "entregado"
+)
+
+type Plato struct {
```
¿Esto es un objeto valor o una entidad?

iscoct replied:
A mi me parece un objeto valor porque aunque platos pueda tener en el futuro pasos, el optimizador de pasos de los platos, no será una lógica del plato en sí, sino de una entidad que optimice los pasos de los platos a hacer.

iscoct replied:
¿Cómo lo ves tú? @JJ

JJ replied:
> A mi me parece un objeto valor porque aunque platos pueda tener en el futuro pasos, el optimizador de pasos de los platos, no será una lógica del plato en sí, sino de una entidad que optimice los pasos de los platos a hacer.

Si pides comidas y bebidas y tienes que optimizar "la comanda", ¿por qué son historias de usuario diferentes?

iscoct replied:
Puedo unirlo en una sola @JJ y que sean 2 issues: optimizar el tiempo en sacar los platos y optimizar el tiempo necesario para sacar las bebidas.

Lo distingo porque aunque el fin es optimizar el tiempo, puedo enfocarme en optimizar el tiempo en sacar las bebidas o el tiempo en sacar los platos, o hacer ambos.

Se optimiza el tiempo en sacar las bebidas mejorando la comunicación y la información que se muestra a los empleados del restaurante y los platos haciendo los mismo (mejorando comunicación) y aparte la posibilidad de optimizar los pasos a hacer entre los distintos platos entre todas las comandas.

El optimizador de platos creo que debería ser otra entidad distinta a la de los platos, las bebidas y las comandas, la cual se definiría en un futuro.
En los primeros milestone quiero enfocarme en optimizar la comunicación.

JJ commented on types/models/comanda.go:
```diff
@@ -0,0 +1,10 @@
+type Comanda struct {
+	Mesa      Mesa
+	Timestamp time.Time
+	Platos    []Plato
```
Las historias de usuario de bebidas y platos son diferentes. Una es #13 y otra es #5 (entiendo) ¿Qué HUs hay en esta historia de usuario? ¿Qué issue ha planteado el problema si las bebidas van en la comanda o no? ¿Cómo usa la comanda la lógica de negocio? ¿#13 y #15 usan lógicas de negocio diferentes?

iscoct replied:
Lo dicho [aquí](https://github.com/iscoct/cotan/issues/27) está relacionada con ambas historias porque en una comanda se pide comidas y bebidas.

El problema está definido en [este issue](https://github.com/iscoct/cotan/issues/21) y en [este](https://github.com/iscoct/cotan/issues/15).

 Me parece que son distintas lógicas de negocio en el sentido de que la persona que entrega los platos puede no ser la misma que entrega las bebidas, es una manera de separar responsabilidades aunque una persona pueda hacer ambas al fin y al cabo.

Ya me dice cómo lo ves @JJ

JJ replied:
Yo no tengo que verlo de ninguna forma, es @jmramirezG la persona que tiene que proponer una solución que sirva a la lógica de negocio.

JJ reviewed (changes requested):
Ha mejorado bastante el diseño, pero convendría que realmente mirarais DDD para crear un buen diseño de estructuras de datos, mediante issues que planteen problemas

jmramirezG committed:
Advance #19 as we add support to compare 2 Comandas based on their timestamp

jmramirezG committed:
Advance #19 as we add data estructures to store Comandas (in order) and Empleados (not ordered)

jmramirezG committed:
Advance #19 as we add data estructures to store Comandas (in order) and Empleados (not ordered)

jmramirezG commented:
Hola @JJ, con los cambios que acabo de hacer se puede guardar de forma ordenada según el Timestamp las diferentes comandas, que es justo lo que se describe en los requisitos del [M1](https://github.com/iscoct/cotan/milestone/4), en el que se indica que las comandas han de estar ordenadas.

Faltaría algo más para dar soporte a la lógica de negocio propuesta?

jmramirezG committed:
Advance #19 as we add a method to show current status of the different kinds of food in the current Comanda

jmramirezG commented:
Añado también para poder consultar el estado de los platos de una comanda, ajustándonos de nuevo a lo especificado en el M1

jmramirezG committed:
close #19 because it is not related to business logic

close #23 as it is related to #19, which was discarded
close #24 as it is related to #19, which was discarded
close #26 as it is related to #19, which was discarded

jmramirezG committed:
close #33 as we add the datatypes discussed there

close #27 as we create the Comanda datatype
close #25 as we create the Plato datatype
create iv.yaml file

iscoct commented on iv.yaml:
```diff
@@ -0,0 +1,4 @@
+lenguaje: go
+entidad: 
+  - types/models/plato.go
```
Plato no es una entidad creo tal y como está definida:
- 2 platos iguales pueden prepararse y terminarse a la vez.

Ejemplo:
- Macarrones para una mesa, 3 platos, una misma olla, se preparan todos a la vez y después se sacan 3 raciones de un plato mayor hecho.

jmramirezG replied:
Solucionado!

iscoct commented:
go.mod falta y la PR tiene conflictos.

iscoct commented on types/models/instrumento.go:
```diff
@@ -0,0 +1,5 @@
+package models
+
+type Instrumento struct {
```
Tal y como has diseñado instrumento y paso, no puedo saber cuándo un instrumento está ocupado o no si no me equivoco.

jmramirezG replied:
Se puede inferir cuando un instrumento está en uso de la siguiente manera.

Si un paso está en preparación y requiere un instrumento, ese instrumento está en uso, en caso contrario, ese instrumento no está siendo usado por ese paso.

Para saber si un paso está en preparación, tan solo tenemos que fijarnos en si tiene tiempo de inicio o no.

iscoct replied:
Suena bien, sí.

Guay entonces.

iscoct commented on types/models/paso.go:
```diff
@@ -0,0 +1,13 @@
+type PasoElaboracion struct {
+	Nombre 			string
+	Instrumento 	Instrumento
+	Duracion		int
```
¿Este int que representa exactamente?

jmramirezG replied:
Ese int sería la duración en segundos del paso de elaboración.

iscoct replied:
Okay

jmramirezG commented:
> go.mod falta y la PR tiene conflictos.

El archivo de go.mod aparece en el PR.

Voy a revisar los conflictos

jmramirezG committed:
remove plato from iv.yaml, as it is not an entity

jmramirezG reviewed (commented):
En general creo que esto se ajusta mejor a la lógica de negocio de tu problema, si quieres que realice algún otro cambio, no dudes en ponerlo por aquí!

jmramirezG commented:
> > Los conflictos han sido por la mezcla de código de tu nuevo main con este al mergear el Objetivo 3, como no tengo acceso de escritura al repositorio tienes que resolverlos tu.
> > Respecto a instrumento, es prácticamente similar.
> > El iv.yaml ahora también es igual, sólo ha cambiado la representación, yo lo he puesto en la misma línea, tu en la de abajo.
> > Los README están cambiados de sitio, ya eliges el que quieres conservar en docs y el que quieres conservar como principal.
> 
> No tiene nada que ver que no tengas acceso de escritura, deberías de descargarte los cambios en tu main en el fork y hacer el merge en tu rama del fork para subirlos otra vez.

Perfecto! Hago eso entonces.

jmramirezG commented:
He intentado hacer un rebase pero ha reventado un poco, por eso salen todos los commits otra vez.

No obstante, ya debería estar!

Espero feedback.

iscoct commented:
> He intentado hacer un rebase pero ha reventado un poco, por eso salen todos los commits otra vez.
> 
> No obstante, ya debería estar!
> 
> Espero feedback.

El rebase no hacía falta, ahora hay 50 commits para la PR jajajajaj

jmramirezG commented:
Solicitamos revisión @JJ

iscoct committed:
Eliminada logica innecesaria

Un commit de Jose fui incluido sin intención de hacerlo.
Eliminado logica de la lista de empleados y comandas

iscoct committed:
Added again campo automatizar en iv.yaml

iscoct committed:
Added salto de linea en iv.yaml

JJ reviewed (approved):
:+1:
